awkward/array/jagged.py

```diff
@@ -41,21 +40,23 @@ def counts2offsets(cls, counts):
 
     @classmethod
     def offsets2parents(cls, offsets):
-        out = cls.numpy.zeros(offsets[-1], dtype=cls.JaggedArray.fget(None).INDEXTYPE)
-        cls.numpy.add.at(out, offsets[offsets != offsets[-1]][1:], 1)
-        cls.numpy.cumsum(out, out=out)
-        if offsets[0] > 0:
-            out[:offsets[0]] = -1
-        return out
+        dtype = cls.JaggedArray.fget(None).INDEXTYPE
+        counts = cls.numpy.diff(cls.numpy.r_[dtype.type(0), offsets])  # numpy >= 1.16: diff prepend
+        indices = cls.numpy.arange(-1, len(offsets) - 1, dtype=dtype)
+        return cls.numpy.repeat(indices, awkward.util.windows_safe(counts))
 
     @classmethod
     def startsstops2parents(cls, starts, stops):
-        out = cls.numpy.full(stops.max(), -1, dtype=cls.JaggedArray.fget(None).INDEXTYPE)
-        lenstarts = len(starts)
-        i = 0
-        while i < lenstarts:
-            out[starts[i]:stops[i]] = i
-            i += 1
+        dtype = cls.JaggedArray.fget(None).INDEXTYPE
+        out = cls.numpy.full(stops.max(), -1, dtype=dtype)
+        indices = cls.numpy.arange(len(starts), dtype=dtype)
+        counts = stops - starts
+        parents = cls.numpy.repeat(indices, awkward.util.windows_safe(counts))
+        contiguous_offsets = cls.JaggedArray.fget(None).counts2offsets(counts)
+        content_indices = cls.numpy.arange(contiguous_offsets[-1])
+        content_indices -= cls.numpy.repeat(contiguous_offsets[:-1], awkward.util.windows_safe(counts))
+        content_indices += starts[parents]
+        cls.numpy.put(out, awkward.util.windows_safe(content_indices), parents)
```

Comment:
Anything that gets rid of that Pythonic `while` loop is an improvement. Thank you!

Reply:
Just a comment; nothing to do here.